wordpress-mobile/WordPress-iOS
# Cancel media auto-uploads after 3 failures

Moves forward #12369 for Media.

This PR makes sure that media uploads are not retried after the 3rd automatic retry fails.

PS: @shiki since we have model 90 in our feature branch, this PR will most likely collide - unfortunately there's no easy way for me to make CD changes, target master, and not break our feature branch.

## Important Notes:

- This PR is purposedly aimed at `develop` since the changes should be fairly safe, and also valuable to reduce the impact of some recent issues with failing uploads.
- We'll need to resolve CoreData conflicts when merging this into our feature branch.  That's unfortunately unavoidable on my end.
- The PR is kind of simple, but it took me a while to wrap it up.  There were several small things that made the adventure more complicated than it should have been.  For this reason, this PR has no automated testing now, and I'll analyze tomorrow if I can add some.  Please though, go ahead and start reviewing the PR.

## Testing:

### Preparation

(A) Plase a breakpoint here: https://github.com/wordpress-mobile/WordPress-iOS/blob/ea2c870222e723512dd5d19315fa3a3e1d2fb8a1/WordPress/Classes/Services/MediaService.m#L258

(B) Plase a breakpoint here: https://github.com/wordpress-mobile/WordPress-iOS/blob/3dc497cb38038d664c13dd8994ebe4b3361c39a8/WordPress/Classes/Services/MediaService.m#L226

Make sure you have no media pending auto-upload.  This can be tricky to verify, but considering the logic introduced in this PR, it should be enough to run & kill the App a few times while online to resolve (zero out) all pending media uploads.

### Test 1: Check that a successful manual upload doesn't increase the failure counter.

1. Create a new post
2. Insert an image
3. Make sure that breakpoint (A) is hit as soon as the upload starts.

### Test 2: Check that a failed manual upload leaves the failure counter at 1.

1. Be offline
2. Create a new post
3. Insert an image
4. Make sure that breakpoint (A) is hit as soon as the upload starts.
5. Make sure that breakpoint (B) is NOT hit when the upload fails (can take a while).

### Test 3: Check that manual retries reset the failure counter.

1. Do all the steps from test 2.
2. When the notice with the error comes up, tap "Retry".
3. Make sure that breakpoint (A) is hit as soon as the upload starts.
4. Make sure that breakpoint (B) is NOT hit when the upload fails (can take a while).

### Test 4: Check that automatic retries further increment the counter.

1. Do all the steps from test 2.
2. Stop the app and relaunch it with Xcode (still being offline).
3. Make sure (A) is not hit, but (B) is, and the counter increments to 2.
4. Stop the app and relaunch it with Xcode (still being offline).
5. Make sure (A) is not hit, but (B) is, and the counter increments to 3.
6. Stop the app and relaunch it with Xcode (still being offline).
7. Make sure (A) is not hit, but (B) is, and the counter increments to 4.
8. Stop the app and relaunch it with Xcode (still being offline).
9. Make sure no other breakpoints are hit.

### Test 5:

1. For this test, remove the breakpoints.
2. Do all steps from test 4.
3. Tap "Retry" on the failed post.
4. Make sure the post is uploaded just fine.

## Update release notes:


## Tasks 

- [x] RFC on https://github.com/wordpress-mobile/WordPress-iOS/pull/12416#issuecomment-526238430
- [x] RFC on https://github.com/wordpress-mobile/WordPress-iOS/pull/12416#pullrequestreview-281642098
